### DISCUSSION TITLE
Need help with conditional workflows using paths

### DISCUSSION THREAD

**Discussion by y4m4p** [community/community#25423] [Actions]
<p>Hello, I am currently trying to build a workflow that enables Continuous Deployment for an web application.</p>
<p>I am currently creating 2 different docker images, one for Nginx, and another for the application.</p>
<p>Due to the fact that the Nginx docker image doesn’t change that often, but application image does, I am trying to make the Nginx deployment workflow run ONLY when certain files under “nginx” directory has changed.</p>
<hr>
<p>First, I tried with the next configuration to recreate this manner, but due to the fact that “ignore-paths” only ignores the workflow IFF the changed file is under “nginx” directory, changing other files which is not in the “nginx” directory files would still trigger the Application workflow, running both workflows which is… quite not as I expected.</p>
<h1>Applciation workflow yml</h1>
<pre><code>name: Automatic Application Deployment to Production

on:
push:
branches:
- master
ignore-paths:
- 'nginx/**'

jobs:
build-and-deploy:
name: build and deploy to production
...(build image for application, and push to ECR, and deploy to ECS)
</code></pre>
<h1>Nginx workflow yml</h1>
<pre><code>name: Automatic Nginx and Application Deployment to Production

on:
push:
branches:
- master
paths:
- 'nginx/**'

jobs:
build-and-deploy:
name: build and deploy to production
...(build image for Nginx and application, and push to ECR, and deploy to ECS)
</code></pre>
<hr>
<p>Another thing I tried, is to mash both the Nginx and Application workflow into one yml file, and try to conditionally change the workflow to run, or not to run the steps within the jobs like the following.</p>
<p>But due to the fact that there is currently no way to condition the steps using the file changes, I am currently stuck with how to go on with this method.</p>
<h1>Application and Nginx workflow yml</h1>
<pre><code>name: Automatic Nginx and Application Deployment to Production

on:
  push:
    branches:
      - master

jobs:
  build-and-deploy:
 name: Build application image and deploy to production
 runs-on: ubuntu-latest
 steps:
-uses: actions/checkout@master
 with:
 ref: ${{ github.ref }}
-name: Configure for AWS deployment...
 ...  
 - name: Build and push Nginx image to ECR  
 if: \*\*SOME WAY TO DETECT THE FILE CHANGE\*\* \&lt;==== NEED HELP HERE...  
 ...  
 - name: Build and push application image to ECR, and Deploy to production  
 ...
</code></pre>
<p>I’m hoping to gather any information, or any third-party actions that enables me to run this conditional steps.</p>
<p>Any information, suggestion, critisism is welcome. </p>
<p>Thank you in advance.</p>

**Comment by chrispat** [comment]
<p>Your first two workflows look right to me.  The first one will build based on any push that includes changes that are not under the Nginx path but if the push only changes the Nginx path it will not run.  The second should run only if the changes are under the Nginx path.  </p>
<p>Is that not the behavior you are seeing?</p>

**Comment by y4m4p** [comment]
<p>chirspat,</p>
<p>Greatly appreciate your reply.</p>
<blockquote>
<p><a class="mention" href="/chrispat">@chrispat</a> wrote:</p>
</blockquote>
<blockquote>
<p>Your first two workflows look right to me.  The first one will build based on any push that includes changes that are not under the Nginx path but if the push only changes the Nginx path it will not run.  The second should run only if the changes are under the Nginx path.  </p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<blockquote>
<p>Is that not the behavior you are seeing?</p>
</blockquote>
<p>Yes, what you are stating is correct and it is the behaviour I am seeing, but this was not the behaviour I wanted…</p>
<p>I am very sorry that I didn’t mention another constraint in the first post, but this is how I wanted things to go.</p>
<h3>Constraints</h3>
<p>Due to how application container depends on the Nginx configuration, I MUST sequentially run the deployment like the following.</p>
<ul>
<li>Nginx build and deploy =&gt; Application build and deploy</li>
</ul>
<h3>Rules of workflow</h3>
<ol>
<li>
<p>If any of the files under “nginx” changes, it runs the Nginx workflow, but not the Application workflow </p>
</li>
<li>
<p>If any of the files under “nginx” does not change, it does not run the Nginx workflow, but instead runs the Application workflow</p>
</li>
<li>
<p>If multiple files change INCLUDING the files under “nginx” directory, it runs the Nginx workflow, but it does not run the Application workflow.</p>
</li>
<li>
<p>If mutliple files change other than the files under “nginx” directory, it does not run the Nginx workflow, but instead runs the Application workflow</p>
</li>
</ol>
<h3>Workflow configuration</h3>
<ul>
<li>Nginx workflow =&gt; build and push “Nginx image” AND build and push “Application image” to ECR, and deploys to ECS</li>
<li>Applciation workflow =&gt; build and push “Applciation image” to ECR, and deploys to ECS</li>
</ul>
<p>So, yes <code>paths</code> and <code>ignore-paths</code> works for 1 and 2 for this case.</p>
<p>But for 3 and 4, specifying <code>paths</code> and <code>ignrore-paths</code> wouldn’t work, because it runs both workflows due to its conditioning nature.</p>
<p>I thought of a way to go about this for 2 ways, but I couldn’t find a way to materialize these thoughts.</p>
<p>A. If there was a way to configure (order) the workflows sequentially, so that the Nginx workflow always starts and ends before the Application workflow, then I could be able to run Nginx workflow =&gt; Application workflow consistantly. (even though, 2 application deployment runs… it gets the job done.)</p>
<p>B. The second yml file I presented in the first post, which enables me to conditionally run a step when a file change is detected. This enables me to run the same workflow everytime, and consistantly run the steps in order.</p>
<pre><code>jobs:
  build-and-deploy:
 name: Build application image and deploy to production
 runs-on: ubuntu-latest
 steps:
-uses: actions/checkout@master
 with:
 ref: ${{ github.ref }}
-name: Configure for AWS deployment...
 ...  
 - name: Build and push Nginx image to ECR  
 if: \*\*SOME WAY TO DETECT THE FILE CHANGE\*\* \&lt;==== NEED HELP HERE...  
 ...  
 - name: Build and push application image to ECR, and Deploy to production  
 ...
</code></pre>
<p>I would like to know if there are any ways to materialize the above thoughts.</p>
<p>Thank you for taking your time with me.</p>

**Comment by y4m4p** [comment, marked as answer]
<p>I figured it out…</p>
<p>Writing with <code>ignore-path</code> will not run the workflow if and only if the changed files are under those certain paths.</p>
<pre><code>on:
  push:
    branches:
      - master
    ignore-paths:
      - 'nginx/**'
</code></pre>
<p>But, by writing the path using <code>!</code> like below, this will never run the workflow when there are changes under that specified path.<br>
This enables me to condition the workflow which will never run, even when there are other file changes.</p>
<p>This was the behaviour I wanted.</p>
<pre><code>on:
  push:
    branches:
      - master
    paths:
      - '!nginx/**'
</code></pre>
<p>Thank you for your help, chirspat :slight_smile:</p>

**Comment by munrocket** [comment]
<p>Is there another way based on context analysis?</p>

**Comment by soffes** [comment]
<p>This should be <code>paths-ignore</code> instead.</p>

**Comment by dorny** [comment]
<aside class="quote no-group" data-username="y4m4p" data-post="1" data-topic="16650">
<div class="title">
<div class="quote-controls"></div>
 y4m4p:</div>
<blockquote>
<p><code>if: \*\*SOME WAY TO DETECT THE FILE CHANGE\*\* \&lt;==== NEED HELP HERE...</code></p>
</blockquote>
</aside>
<aside class="quote no-group" data-username="munrocket" data-post="5" data-topic="16650" data-full="true">
<div class="title">
<div class="quote-controls"></div>
 munrocket:</div>
<blockquote>
<p>Is there another way based on context analysis?</p>
</blockquote>
</aside>
<p>You could use <strong><a href="https://github.com/dorny/paths-filter" rel="noopener nofollow ugc">paths-filter</a></strong> - it’s a 3rd party Github action created exactly for this purpose.<br>
There’s quite self explanatory <a href="https://github.com/dorny/paths-filter#example" rel="noopener nofollow ugc">example</a> in the README.</p>
<p>Workflows can’t be composed into bigger units and it’s not possible to specify order in which they will run. If you have conditional steps  - based on which files were modified, you can either use <code>paths</code> + <code>paths-ignore</code> and create  workflow for each possible combination of the steps or use the <code>if</code> clause with action to detect the modified files.</p>
<p>Having more workflows has a benefit of using just a built-in functionality.<br>
However it wouldn’t work well for more complex scenarios.<br>
With more components, you could end up with a dozen of workflows where some parts would have to be duplicated many times.</p>
<p>Using <code>if</code> clauses within single workflow job has also another benefit - all artifacts of previous steps are available without need to upload them somewhere.</p>

**Comment by fkirc** [comment]
<p>I published this Action that implements an improved “paths-ignore” : <a href="https://github.com/fkirc/skip-duplicate-actions" rel="noopener nofollow ugc">https://github.com/fkirc/skip-duplicate-actions</a></p>
<p>It is basically a smarter path-filter that can be used anywhere within a workflow.<br>
You may give it a try if GitHub’s “path-ignore” isn’t good enough.</p>

**Reply by y4m4p** [reply]
Hey, sorry to reply on a 3 year necro'd topic, but thanks for this implementation and notice.
Looks really good, I'll give this a try on the next application development. 😃
